src/smartcontracts/smartContract.ts

```diff
@@ -110,27 +115,58 @@ export class SmartContract implements ISmartContract {
     deploy({ deployer, code, codeMetadata, initArguments, value, gasLimit, gasPrice, chainID }: DeployArguments): Transaction {
         Compatibility.guardAddressIsSetAndNonZero(deployer, "'deployer' of SmartContract.deploy()", "pass the actual address to deploy()");
 
-        codeMetadata = codeMetadata || new CodeMetadata();
-        initArguments = initArguments || [];
-        value = value || 0;
+        const config = new TransactionIntentsFactoryConfig(chainID.valueOf());
+        const scIntentFactory = new SmartContractTransactionIntentsFactory({
+            config: config,
+            abi: this.abi
+        });
 
-        let payload = new ContractDeployPayloadBuilder()
-            .setCode(code)
-            .setCodeMetadata(codeMetadata)
-            .setInitArgs(initArguments)
-            .build();
+        const bytecode = Uint8Array.from(Buffer.from(code.toString(), 'hex'));
+        const metadataAsJson = this.getMetadataPropertiesAsObject(codeMetadata);
 
-        let transaction = new Transaction({
-            receiver: Address.Zero(),
+        const intent = scIntentFactory.createTransactionIntentForDeploy({
             sender: deployer,
+            bytecode: bytecode,
+            gasLimit: gasLimit.valueOf(),
+            args: initArguments,
+            isUpgradeable: metadataAsJson.upgradeable,
+            isReadable: metadataAsJson.readable,
+            isPayable: metadataAsJson.payable,
+            isPayableBySmartContract: metadataAsJson.payableBySc
+        });
+
+        return new Transaction({
+            receiver: Address.fromBech32(intent.receiver),
+            sender: Address.fromBech32(intent.sender),
             value: value,
-            gasLimit: gasLimit,
+            gasLimit: new BigNumber(intent.gasLimit).toNumber(),
             gasPrice: gasPrice,
-            data: payload,
+            data: new TransactionPayload(Buffer.from(intent.data!)),
             chainID: chainID
         });
+    }
 
-        return transaction;
+    private getMetadataPropertiesAsObject(codeMetadata?: ICodeMetadata): {
+        upgradeable: boolean,
+        readable: boolean,
+        payable: boolean,
+        payableBySc: boolean
+    } {
+        let metadata: CodeMetadata;
+        if (codeMetadata) {
+            metadata = CodeMetadata.fromBytes(Buffer.from(codeMetadata.toString(), "hex"));
+        }
+        else {
+            metadata = new CodeMetadata();
+        }
+        const metadataAsJson = metadata.toJSON() as {
```

Comment:
Alternatively, we can directly make the fields of `CodeMetadata` as `public readonly`:

https://github.com/multiversx/mx-sdk-js-core/blob/main/src/smartcontracts/codeMetadata.ts

Then use those, without an extra mapping to JSON.

Reply:
Not really good if someone uses the `toggle...` methods.